[MRG] Add verbose option for Pipeline and Feature Union.

kdexd opened the pull request:
#### Reference Issue
Fixes #5298, #5321

#### What does this implement/fix? Explain your changes.

This PR adds a new argument `verbose` to `Pipeline` and `FeatureUnion`, which is False by default. Setting this `True` will print useful information during execution. Better to be explained by a code snippet:

**Common code snippet (imports and data loading):**
```python
from sklearn.datasets import load_iris
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline, FeatureUnion
from sklearn.preprocessing import Imputer

data = load_iris(return_X_y=True)
```

**Pipeline Snippet:**
```python
p = Pipeline([('imputer', Imputer()), ('linreg', LinearRegression())],
             verbose=True)

p.fit(data[0], data[1])
y_pred = p.predict(data[0][0].reshape(1, -1))
print("Predicted: %f" % y_pred)
```

**Expected Output:**
```
[Pipeline] (step 1 of 2) imputer ............................ 0.00053s
[Pipeline] (step 2 of 2) linreg ............................. 0.00043s
[Pipeline] Total time elapsed: .............................. 0.00096s
```

**FeatureUnion Snippet:**
```python
f = FeatureUnion([('imputer', StandardScaler()),
                  ('pca', PCA(n_components=2))], verbose=True)

f.fit(data[0], data[1])
data_t = f.transform(data[0][0].reshape(1, -1))
print("Transformed: %r" % data_t)
```

**Expected Output:**
```
[FeatureUnion] (step 1 of 2) imputer ........................ 0.00014s
[FeatureUnion] (step 2 of 2) pca ............................ 0.00032s
[FeatureUnion] Total time elapsed: .......................... 0.00065s
Transformed: array([[-0.90068117,  1.03205722, -1.3412724 , -1.31297673, -2.68420713,
         0.32660731]])
```

#### Any other comments?

* TODO (will keep on editing this description):
- [x] Initial PR work, add verbosity to Pipeline.
- [x] Add tests for verbosity of Pipeline.
- [x] Add verbosity to FeatureUnion.
- [x] Add tests for verbosity to FeatureUnion.
- [x] Modify both methods to print in such a way that all lines have 70 characters.
- [x] Add _CHANGELOG_ entry.

I had open PRs else where, so it took me a while to get back to this. Thanks @jnothman for mentioning me on issue thread.

Commit:
ENH Add verbose option and corresponding tests for Pipeline.

karandesai-96 commented on sklearn/pipeline.py:
```diff
@@ -337,13 +362,21 @@ def fit_transform(self, X, y=None, **fit_params):
-            return last_step.fit(Xt, y, **fit_params).transform(Xt)
+            Xt = last_step.fit(Xt, y, **fit_params).transform(Xt)
+        if self.verbose:
+            self._print_final_step(final_step_start_time, time_elapsed_so_far)
```
I would have included `if self.verbose` condition in the method itself, to avoid repeated checks at every calls, but this way it felt semantically more appropriate - "if verbose then print final step".

Else simply the call would state "print final step" while it internally wouldn't have printed if not verbose. I hope that is fine ?

Commit:
ENH Add verbose option and corresponding tests for FeatureUnion.

karandesai-96 commented:
`Pipeline` performs stepwise operations sequentially, while `FeatureUnion` executes parallel jobs for transformers. So for `Pipeline`, the total time elapsed will be sum of individual times taken at each step (_and it is implemented that way too_). On the other hand, total time elapsed for a `FeatureUnion` varies on the number of jobs, hence it is separately monitored, and so it necessarily need not be summation of individual times taken at each step..

Commit:
ENH Add _pretty_print method to print from Pipeline properly.

- Each line printed by Pipeline and FeatureUnion, when their
  verbosity mode is on, will be 70 characters long.

Commit:
Add a changelog entry about verbosity of Pipeline.

karandesai-96 commented:
/ping: @jnothman @amueller @raghavrv :rocket:

raghavrv commented:
This is a neat work... Thanks!

karandesai-96 commented:
Thanks @raghavrv .
Also, this PR if merged, should close #5321

karandesai-96 commented on doc/modules/pipeline.rst:
```diff
@@ -148,7 +149,7 @@ object::
     Pipeline(...,
              steps=[('reduce_dim', PCA(copy=True,...)),
-                    ('clf', SVC(C=1.0,...))])
+                    ('clf', SVC(C=1.0,...))], verbose=False)
```
Well then i'll wait for others' suggestion meanwhile. I'm not sure which should be preferred either.

karandesai-96 replied:
If we set an analogy, `transformer_weights` are shown here in same documentation. They're not having any value in this particular example.
```python
    >>> combined # doctest: +NORMALIZE_WHITESPACE, +ELLIPSIS
    FeatureUnion(n_jobs=1,
                 transformer_list=[('linear_pca', PCA(copy=True,...)),
                                   ('kernel_pca', KernelPCA(alpha=1.0,...))],
                 transformer_weights=None, verbose=False)
```

We can use this as a tie breaker :laughing:

karandesai-96 commented on doc/whats_new.rst:
```diff
@@ -53,6 +53,10 @@ New features
 
+   - Added optional parameter ``verbose`` in :class:`pipeline.Pipeline` and
+     :class:`pipeline.FeatureUnion` for showing progress and timing of each
+     step. :issue:`8568` by :user:`Karan Desai <karandesai-96>`.
```
Done :white_check_mark:

karandesai-96 commented:
/ping: @jnothman Resolved conflicts with master which moved forward after I opened my PR.

glemaitre commented:
@karandesai-96 You should not merge master into your branch but rebase your branch to master.
Revert to the previous commit and rebase.

karandesai-96 commented:
@glemaitre I used the web interface for a very small conflict resolving, tried it for the first time. Agreed with you, not impressive.

Reverting, rebasing.

karandesai-96 commented:
Rebased.

karandesai-96 commented:
Rebased.

jnothman commented:
i often merge master into my branch. usually works just fine. I'll try look
at this soon

On 10 Apr 2017 3:10 pm, "Karan Desai" <notifications@github.com> wrote:

> Rebased.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8568#issuecomment-292851126>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz67jDPLMIYHq-FvNnJaRYmbQzNPP6ks5rubmogaJpZM4MY6aU>
> .
>

glemaitre commented:
I cannot say that this is a good practice or the right way to do it. I
just got it while contributing in scikit-image in which I did merge and
have been forced to reset and rebase:
http://scikit-image.org/docs/0.11.x/contribute.html#divergence-between-upstream-master-and-your-feature-branch

It seems that you have the same in the guide in scikit-learn:
http://scikit-learn.org/stable/developers/contributing.html#how-to-contribute

From the top of the head, I was thinking that the reasons were: merging
make the history difficult to read but rebasing rewrite the history.
--
Guillaume Lemaitre
INRIA Saclay - Parietal team
Center for Data Science Paris-Saclay
https://glemaitre.github.io/

jnothman commented:
I think Github deals with merging in master better than it used to. I no
longer recommend rebases: that requires resolving conflicts on every commit
rather than just between the heads (and squash-merging the final PR can
handle whatever you did along the way as it'll just diff the final heads as
long as they are merge-safe). It also messes up historical commit
references and requires force pushes and other Dangerous Ideas.

On 12 April 2017 at 21:29, Guillaume Lemaitre <notifications@github.com>
wrote:

> I cannot say that this is a good practice or the right way to do it. I
> just got it while contributing in scikit-image in which I did merge and
> have been forced to reset and rebase:
> http://scikit-image.org/docs/0.11.x/contribute.html#
> divergence-between-upstream-master-and-your-feature-branch
>
> It seems that you have the same in the guide in scikit-learn:
> http://scikit-learn.org/stable/developers/contributing.html#how-to-
> contribute
>
> From the top of the head, I was thinking that the reasons were: merging
> make the history difficult to read but rebasing rewrite the history.
> --
> Guillaume Lemaitre
> INRIA Saclay - Parietal team
> Center for Data Science Paris-Saclay
> https://glemaitre.github.io/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8568#issuecomment-293549319>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wjMpKHgN5NOr0o9iujLHWDvRSa6ks5rvLW0gaJpZM4MY6aU>
> .
>

glemaitre commented:
I completely agree with the issue linked to resolving every conflicts,
especially when resurrecting old PRs.

In fact, I personally do not have a strong opinion regarding the policy
to follow. Merge has only 1 con regarding the readability which is
partially solved when we are actually squashing/merge at the end.

Do you think that we should change the contributing guide in this regard
if there is consensus behind a only merging policy?

--
Guillaume Lemaitre
INRIA Saclay - Parietal team
Center for Data Science Paris-Saclay
https://glemaitre.github.io/

jnothman commented:
I'd be in favour of updating the docs, yes

On 12 Apr 2017 10:31 pm, "Guillaume Lemaitre" <notifications@github.com>
wrote:

>
> I completely agree with the issue linked to resolving every conflicts,
> especially when resurrecting old PRs.
>
> In fact, I personally do not have a strong opinion regarding the policy
> to follow. Merge has only 1 con regarding the readability which is
> partially solved when we are actually squashing/merge at the end.
>
> Do you think that we should change the contributing guide in this regard
> if there is consensus behind a only merging policy?
>
> --
> Guillaume Lemaitre
> INRIA Saclay - Parietal team
> Center for Data Science Paris-Saclay
> https://glemaitre.github.io/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8568#issuecomment-293561431>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wiC4PMDDhZmw4c8nRaBj9sEilLdks5rvMQWgaJpZM4MY6aU>
> .
>

jnothman commented on sklearn/pipeline.py:
```diff
@@ -25,7 +26,24 @@
+
+    Parameters
+    ----------
+    step_info : str
```
Why is this preferred over `description`, `elapsed` or something else that further ensures consistency?

I would like to see `logger.short_format_time(elapsed)` used as in grid search, and this code refactored into utils with that.

amueller replied:
What's the code in grid search you're talking about? The only logging is within joblib, right?

amueller commented:
could you please resolve the conflicts?

amueller reviewed (commented):
I think this would be an awesome feature to have, but it should really live in a logger or something. Right now it's cluttering up the already complicated pipeline code.

karandesai-96 commented:
Hey @amueller, sorry I was away for a while I couldn't respond. I agree with your review. I will resolve the conflicts meanwhile I'll look into a more sane way for logging as you say.

jnothman commented:
The verbosity from grid search lives in model_selection/_validate.py I
think.

On 22 Jul 2017 2:00 pm, "Karan Desai" <notifications@github.com> wrote:

> Hey @amueller <https://github.com/amueller>, sorry I was away for a while
> I couldn't respond. I agree with your review. I will resolve the conflicts
> meanwhile I'll look into a more sane way for logging as you say.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8568#issuecomment-317152384>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61rnEjFevRB91GKuw2gr5qQcQuAYks5sQXPXgaJpZM4MY6aU>
> .
>

amueller commented:
Indeed, and it uses joblib.logger
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/model_selection/_validation.py#L479

jnothman commented:
I'd like to see this progress... Are you continuing to work on it @karandesai-96

EBazarov commented:
**Verbosity in pipeline very helpful feature**, especially for debugging. Any plans to merge this PR ?

amueller commented:
@EBazarov as soon as there is a reasonable implementation.
